client/src/send_and_confirm_transactions_in_parallel.rs

```diff
@@ -0,0 +1,465 @@
+                }
+            } else {
+                tokio::time::sleep(Duration::from_millis(100)).await;
+            }
+            if let Some(max_valid_block_height_in_remaining_transaction) = transaction_map
+                .iter()
+                .map(|x| x.last_valid_blockheight)
+                .max()
+            {
+                max_valid_block_height = max_valid_block_height_in_remaining_transaction;
+            }
+        }
+
+        if let Some(progress_bar) = progress_bar {
+            set_message_for_confirmed_transactions(
+                progress_bar,
+                confirmed_transactions.load(std::sync::atomic::Ordering::Relaxed),
+                total_transactions,
+                Some(current_blockheight.load(Ordering::Relaxed)),
+                max_valid_block_height,
+                "Checking transaction status...",
+            );
+        }
+    }
+}
+
+// This is a new method which will be able to send and confirm a large amount of transactions
+// usually the send_and_confirm_transactions_with_spinner for QUIC clients are optimized for small number of transactions
+// If we use these methods
+pub async fn send_and_confirm_transactions_in_parallel<T: Signers + ?Sized>(
+    rpc_client: Arc<RpcClient>,
+    websocket_url: String,
+    messages: &[Message],
+    signers: &T,
+    resign_txs_count: Option<usize>,
+    with_spinner: bool,
+    send_over_tpu: bool,
+) -> Result<Vec<Option<TransactionError>>> {
+    let tpu_client = if send_over_tpu {
+        let connection_cache = ConnectionCache::new_quic("connection_cache_cli_program_quic", 1);
+        if let ConnectionCache::Quic(cache) = connection_cache {
+            Some(
+                TpuClient::new_with_connection_cache(
+                    rpc_client.clone(),
+                    &websocket_url,
+                    TpuClientConfig::default(),
+                    cache,
+                )
+                .await?,
+            )
+        } else {
+            // should not ever happen
+            panic!("Expected a quic connection cache");
+        }
+    } else {
+        None
+    };
+    // get current blockhash and corresponding last valid block height
+    let blockhash_and_valid_height = rpc_client
+        .get_latest_blockhash_with_commitment(rpc_client.commitment())
+        .await?;
+    let blockhash_data_rw = Arc::new(RwLock::new(BlockHashData {
+        blockhash: blockhash_and_valid_height.0,
+        last_valid_blockheight: blockhash_and_valid_height.1,
+    }));
+
+    // check if all the messages are signable by the signer
+    let signature_results = messages.iter().map(|x| {
+        let mut transaction = Transaction::new_unsigned(x.clone());
+        transaction.try_sign(signers, blockhash_and_valid_height.0)
+    });
+
+    for signature_result in signature_results {
+        if let Err(e) = signature_result {
+            return Err(e.into());
```

Comment:
prefer `T::from(var)` over `var.into::<T>()` for the sake of self-documentation

Reply:
pretty sure `collect()` is smart enough to transpose the `Vec<Result<T, E>>` into `Result<Vec<T>, E>`, returning the first error. so i think you can just combine this block with the one above and let `?` do the error conversion behind the scenes. bonus, we get lazy eval of the iterator

```rust    
    // check if all the messages are signable by the signer
    messages.iter().map(|x| {
        let mut transaction = Transaction::new_unsigned(x.clone());
        transaction.try_sign(signers, blockhash)
    })
    .collect::<Result<Vec<()>, SignerError>()?;
```